jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpReceiverOverHTTP.java

```diff
@@ -51,21 +53,101 @@ public class HttpReceiverOverHTTP extends HttpReceiver implements HttpParser.Res
     private boolean unsolicited;
     private String method;
     private int status;
+    private Content.Chunk chunk;
 
     public HttpReceiverOverHTTP(HttpChannelOverHTTP channel)
     {
         super(channel);
         HttpClient httpClient = channel.getHttpDestination().getHttpClient();
         parser = new HttpParser(this, -1, httpClient.getHttpCompliance());
         HttpClientTransport transport = httpClient.getTransport();
-        if (transport instanceof HttpClientTransportOverHTTP)
+        if (transport instanceof HttpClientTransportOverHTTP httpTransport)
         {
-            HttpClientTransportOverHTTP httpTransport = (HttpClientTransportOverHTTP)transport;
             parser.setHeaderCacheSize(httpTransport.getHeaderCacheSize());
             parser.setHeaderCacheCaseSensitive(httpTransport.isHeaderCacheCaseSensitive());
         }
+        retainableByteBufferPool = httpClient.getByteBufferPool().asRetainableByteBufferPool();
+    }
 
-        this.retainableByteBufferPool = httpClient.getByteBufferPool().asRetainableByteBufferPool();
+    @Override
+    public void receive()
+    {
+        // This method is the callback of fill interest.
+        // As such, it is called repeatedly until the ContentSourceListener.onContentSource() loop gets started;
+        // meaning firstContent is false and it must register for fill interest if no filling was done
+        // until onContentSource() gets called.
+        // Once onContentSource() gets called, firstContent is true and it must just notify that content may be generated.
+
+        ContentSource contentSource = getContentSource();
+        if (contentSource == null)
```

Comment:
Can this be just `hasContent() => responseState == ResponseState.CONTENT`.

Reply:
Rather `hasContent() => contentSource != null`